common-test-lib/src/main/kotlin/com/google/container/tools/test/ContainerToolsRule.kt

```diff
@@ -64,33 +72,80 @@ class ContainerToolsRule(private val testInstance: Any) : TestRule {
         EdtTestUtil.runInEdtAndWait(ThrowableRunnable { ideaProjectTestFixture.setUp() })
 
         MockKAnnotations.init(testInstance, relaxed = true)
         replaceServices()
+        createTestFiles(description.methodName)
     }
 
     private fun tearDownRule() {
+        filesToDelete.forEach { it.delete() }
         EdtTestUtil.runInEdtAndWait(ThrowableRunnable { ideaProjectTestFixture.tearDown() })
     }
 
     /**
      * Replaces all services annotated with [TestService].
      */
     private fun replaceServices() {
-        for (member in testInstance::class.declaredMemberProperties) {
-            if (member.annotations.filter { it is TestService }.isNotEmpty()) {
-                member as KProperty1<Any?, Any?>
-                member.isAccessible = true
-                val service: Any = member.get(testInstance)!!
-                setService(service)
+        for (member in getMembersWithAnnotation(TestService::class.java)) {
+            member as KProperty1<Any?, Any?>
+            member.isAccessible = true
+            val service: Any = member.get(testInstance)!!
+            setService(service)
+        }
+    }
+
+    /**
+     * Creates all [File] annotated with [TestFile] in the given directory name.
+     *
+     * @param directoryName the name of the directory to create the test files in
+     */
+    private fun createTestFiles(directoryName: String) {
+        for (member in getMembersWithAnnotation(TestFile::class.java)) {
+            member.isAccessible = true
+            if (!member.returnType.isSubtypeOf(File::class.createType()) ||
+                member !is KMutableProperty<*>
+            ) {
+                throw IllegalArgumentException(
+                    "@TestFile can only annotate mutable fields of type " +
+                        "java.io.File"
+                )
+            }
+
+            val annotation: TestFile? = member.findAnnotation()
+            val directory: File = FileUtil.createTempDirectory(directoryName, null)
+            val file = File(directory, annotation?.name)
+            if (!file.createNewFile()) {
+                throw IOException("Can't create file: $file")
             }
+            annotation?.let {
+                if (!it.contents.isEmpty()) {
+                    FileUtil.writeToFile(file, it.contents)
+                }
+            }
+
+            filesToDelete.add(file)
+            member.setter.call(testInstance, file)
         }
     }
 
+    /**
+     * Returns a list of members containing the given annotation.
+     *
+     * @param annotation find members containing this annotation
+     */
+    private fun getMembersWithAnnotation(annotation: Class<out Annotation>):
+        List<KProperty1<out Any, Any?>> =
+        testInstance::class.declaredMemberProperties.filter { member ->
+            member.annotations.filter {
+                annotation.isAssignableFrom(it::class.java)
```

Comment:
Could we use the style used previously, with `it` instead of assignable from and Java class - i.e. in the manner of `(member.annotations.filter { it is annotation }`?

Reply:
so what you are suggesting is actually different behavior.

`annotation.isAssignableFrom(it::class.java)`

that is checking that the current member is a subclass of the passed in `annotation`. The passed in annotation could be `TestService`, or `TestFile`, for example. This allows this method to be reusable.

Does that clear this up or do you still see issues with it?

Reply:
It works, just a bit confusing why we need to support subclassing for annotations, what if we just check for straight class match (i.e. `is` as `instanceof` for Kotlin), won't this be enough for us?

Reply:
hmm, I see what you're saying now. i think you're right. Let me update.

Reply:
aha. now I remember. I tried that myself initially, but the problem is that this code snippet:

`(member.annotations.filter { it is annotation }`

is not allowed. the part after the `is` needs to be a type, not a variable. So as far as I could tell what I have is the only way to do it. If you can think of a better way I'll welcome it.

Reply:
Yeah, it looks like a logical case for a class, but seems not to be working. I'm ok with it in the current form then, maybe we can consult Kotlin reviewers on that as well.

Reply:
I updated it to use `#isInstance` instead of `#isSuperclassOf` to hopefully make it less confusing.

Regarding using a class variable in the `x is y` expression, I think its the same limitation that java imposes on `instanceof`. If you need to use a class instance then you need to do `clazz.instanceOf(foo)`.